src/ports/Jwt.ts

```diff
@@ -0,0 +1,102 @@
+import cookie from 'cookie'
+import { CrossFetchRequest, Fetcher } from 'dcl-catalyst-commons'
+import log4js from 'log4js'
+import ms from 'ms'
+import { generateNonceForChallenge } from '../utils/ProofOfWork'
+
+const LOGGER = log4js.getLogger('JWTPort')
+
+export async function obtainJWT(fetcher: Fetcher, catalystUrl: string): Promise<string | undefined> {
+  try {
+    const response = await fetcher.fetchJson(catalystUrl + '/pow-auth/challenge')
+    const body = JSON.parse(JSON.stringify(response))
+
+    const challenge: string = body.challenge
+    const complexity: number = body.complexity
+    const nonce: string = await generateNonceForChallenge(challenge, complexity)
+
+    const powAuthUrl = new URL('/pow-auth/challenge', catalystUrl).href
+    const challengeBody = JSON.stringify({ challenge: challenge, complexity: complexity, nonce: nonce })
+    const jwtResponse = await fetcher.postForm(powAuthUrl, { body: challengeBody })
+
+    if (!jwtResponse.jwt) {
+      LOGGER.warn('[POW] Could not get a JWT from Pow Auth Server.')
+    }
+    return jwtResponse.jwt
+  } catch (error) {
+    LOGGER.warn(`[POW] Could not get a JWT from Pow Auth Server, due to: ${error}`)
+    return ''
+  }
+}
+
+export function isJWTCookieRemoved(response: Response): boolean {
+  try {
+    const headers = response.headers
+    if (headers) {
+      const setCookie = headers.get('Set-Cookie')
+      if (setCookie && setCookie.includes('JWT=')) {
+        const cookies = cookie.parse(setCookie)
+        return cookies.JWT === ''
+      }
+    }
+    return false
+  } catch {
+    return false
+  }
+}
+
+export function missingJWTInRequest(request: CrossFetchRequest): boolean {
+  const headers: Headers | string[][] | Record<string, string> | undefined = request.requestInit?.headers
+
+  if (!!headers) {
+    if (headers instanceof Headers) {
+      return !hasJWTCookie(headers.get('Cookie') ?? '')
+    } else if (Array.isArray(headers)) {
+      return !headers.find((a) => {
+        a[0] == 'Cookie' && hasJWTCookie(a[1])
+      })
+    } else {
+      return !hasJWTCookie(headers.Cookie ?? '')
+    }
+  } else {
+    return true
+  }
+}
+
+function hasJWTCookie(cookieValue: string): boolean {
+  const cookies = cookie.parse(cookieValue)
+  return cookies.JWT ?? '' !== ''
+}
+
+export function configureJWTMiddlewares(fetcher: Fetcher, baseUrl: string): void {
+  let lastFailedPowEndpointTimestamp: number = 0
+  let minutesToAdd: number = ms('5m')
+  let isRequestingJWT: boolean = false
+  fetcher.overrideDefaults({
+    requestMiddleware: async (request: CrossFetchRequest) => {
+      if (missingJWTInRequest(request) && !isRequestingJWT) {
+        if (lastFailedPowEndpointTimestamp + minutesToAdd < Date.now()) {
+          isRequestingJWT = true
```

Comment:
I would add a try/catch and set `isRequestingJWT` to false on the finally. Because if `obtainJWT` throws an exception, I'm not sure how everything will end up being

Reply:
+1